### ISSUE TITLE
Flag inappropriate comments on challenges

### ISSUE THREAD

**Issue by asimonw** [emfoundation/ce100-app#947]
As a CE100 member, I want to flag irrelevant or offensive comments to the CE100 platform moderators so that I feel safe sharing challenges.

The flag action should invoke a confirmation prompt using the `window.confirm` function.
If the user confirms, an email should be sent to the CE100 team. An admin user can edit or delete the comment if necessary.

The flagged comment will be visible to internal users, primary users of the related org and the poster of the comment. It will be invisible to other users.

**Comment by iteles** [comment]
+ Email goes to CE100 email (automated with comment, who flagged and link to challenge)
+ Highlighted in red for person who flagged the comment
+ Everyone else sees 'This comment has been flagged for moderation'
+ Admin sees red highlight with red 'Needs review' wording + 3 dot menu to edit or delete (not flag)

@fannyBkane @asimonw @georgemillard to confirm at what point the comment is displayed again to the public

**Comment by iteles** [comment]
Estimates include _only_ up to the point where a comment appears as flagged. The resolution portion of this issue for admins is not yet completed, pending @asimonw @fannyBkane & @georgemillard's resolution of the question.

**Comment by asimonw** [comment]
@fannyBkane @georgemillard I suggest that we don't allow the extra flexibility to "edit without resolving" for now. This means that we do one of the following:
+ Option 1: show two options "Resolve" or "Delete". Clicking Resolve would show the comment in edit mode. Saving the edited comment would resolve (i.e. un-flag) the comment as well.
+ Option 2: show the same options as the creator of the comment, "Edit" or "Delete", but then I'd at least name the button "Resolve" rather than "Save" to indicate what the saving the comment will do.

Hope this is clear. Both options only differ in how we name options/buttons, so it should be ok to time estimate I think.

**Comment by fannyBkane** [comment]
@asimonw Cheers for that. 

A couple of thoughts:
In Option 1, 'Delete' also leads to resolving the problem
In Option 2, 'Resolve' doesn't say what is actually going to happen after you click the button

So, I would suggest to go with **Option 2 with a tweak**: after editing, the button would say 'Save and unflag'

**Comment by asimonw** [comment]
@fannyBkane I guess I saw "Resolve" as a positive action, as opposed to "Delete", but indeed if that can cause confusion, happy to go for the more precise "Unflag". @georgemillard @iteles do you think that word is clear enough? I think it might not actually be an official word (but it would only be visible to admins anyway).

@iteles Are you guys happy with my option 2, with @fannyBkane's suggested tweak? Let us know if there are any technical considerations you can think of that we might have missed.

**Comment by georgemillard** [comment]
@asimonw @fannyBkane @iteles 

If an EMF User has Flagged a comment, it may not be clear to the Admin what the exact nature of the problem is. Perhaps allowing the Admin to edit is not necessary for a first iteration. 

If we just gave the Admin the option to Unflag or Delete, would this be a simpler solution for now?

Also, we need to consider emailing the Comment Creator, and the person who flagged the comment (the Flagger!) to inform them of the action taken.

**Comment by asimonw** [comment]
@georgemillard The idea is that the admin (or more likely someone from the team) would have had a conversation with possibly both parties before resolving the comment. Otherwise they wouldn't know how to resolve either way in many cases. If we don't give edit functionality to the admin, no-one would be able to edit it, so it would be an all or nothing situation.

The solution I'm suggesting is almost exactly like the functionality for the owner of the challenge (with an extra unflag trigger), so I don't think it would be much more difficult.

Good point about sending those emails though! That makes sense.

**Comment by iteles** [comment]
We'll be moving onto this in the next couple of few days.

@georgemillard Do you have any thoughts on Alex's thoughts above?
@asimonw Shall we add emails as a separate (but linked) issue and discuss priority for that as an enhancement to this feature?

**Comment by asimonw** [comment]
@iteles I had a chat about this with @georgemillard on Friday, but hadn't updated this issue yet. We'll go for the option 2 with a tweak, as @fannyBkane suggested. I'll add that to the too comment.

I'm happy to create a new, linked issue to handle the emails. We we already had an email as part of the process above (email to C100 when a comment is flagged). Should we group all the email-related bits in the same issue then?

**Comment by iteles** [comment]
@asimonw Thanks for confirming 👍 

If you feel that both are enhancements, then that's a good plan.
However, I think that the first email (to CE100) is a key part of the process as admins have no other way of knowing that something has been flagged. As such, I would keep this email in the current main issue (this one) and add the resolution email as an enhancement in a separate issue which I would imagine is a lower priority.

**Comment by asimonw** [comment]
@iteles That makes a lot of sense. I'll create a new issue for the confirmation emails to challenge owner and commenter, but let's keep the one to notify the admin as part of this issue.

**Comment by Cleop** [comment]
@fannyBkane - please can you share the asset for the flag?

**Comment by Cleop** [comment]
**Users**
-----------
As a **challenge creator** I see:
- an ellipsis on my comments
- a flag icon for others comments
<img src="https://user-images.githubusercontent.com/16775804/31725075-598f2cbe-b41b-11e7-9979-136d74296735.png" width=350px />

As someone who's **commented on someone else's challenge** I see:

- an ellipsis on my comments
- nothing or a flag next to other people's comments (including the challenge creator)
(This looks the same as above)

As an **admin** I see:
- an ellipsis on my comments
- a flag icon and an ellipsis for others comments?

**Once a comment has been flagged:**
-------------
As a **challenge creator** (an org's primary user? or can secondary users create challenges?), **the author of the comment** or **the flagger** I see:
<img src="https://user-images.githubusercontent.com/16775804/31725757-0c02eb1e-b41d-11e7-9733-421b0e904d8c.png" width=350px />
As an **admin** I see:
<img src="https://user-images.githubusercontent.com/16775804/31726191-27ce2be6-b41e-11e7-95f8-92e4a334d895.png" width=350px />
**Note:** the ellipsis instead of flagged wording
As **anyone else** I see:
<img src="https://user-images.githubusercontent.com/16775804/31726130-0543e534-b41e-11e7-9b20-62d11dddaf47.png" width=350px />

**Comment by fannyBkane** [comment]
@Cleop emailed :)

**Comment by iteles** [comment]
Adding an extra 3 hours here to cover the following, with these bullet points also being part of Sprint 9:
+ Unflagging functionality
  + Option 2 here: https://github.com/emfoundation/ce100-app/issues/947#issuecomment-335512903
  + _BUT_ rather than say `Resolve`, the button should read `Save & Unflag`
+ Email to be sent when comment is flagged as per https://github.com/emfoundation/ce100-app/issues/947#issuecomment-336891720
  + This is to be hard-coded to be sent to the ce100 gmail address
  + Proposed text (keeping it simple): `A comment has been flagged by a user of the CE100 membership application. Please follow this link to view the challenge with the flagged comment (which will be highlighted in red) and take action to resolve the issue: (insert URL of challenge with the flagged comment)`

**Comment by asimonw** [comment]
@iteles Would it be possible to have a different hard-coded email value for testing on staging?

**Comment by SimonLab** [comment]
@asimonw yes I can define the email in the environment variable of Heroku like this you will be able to define the email also yourself if you want to change it later on

**Comment by SimonLab** [comment]
The email will be send by the value define with the environment variable ```COMMENT_ADMIN_EMAIL```. This value can be updated in the settings of Heroku, however you might need to restart the application on Heroku

**Comment by asimonw** [comment]
@SimonLab Looks good! Nothing to flag (no pun intended) for now, but I'd like to retest this once #1010 is fixed, to make sure people from other orgs see a "greyed out" box. I was already able to test it for other people of the same org, so I think it will be fine.

**Comment by Cleop** [comment]
@asimonw #1010 is now in `please-test` so feel free to test this too 👍

**Comment by asimonw** [comment]
@Cleop @SimonLab @iteles Now that #1010 is in place, I notice that anyone can flag a comment, while my understanding was that only the owner of the challenge would be able to do that. We can interpret "owner" to be anyone of the organisation the challenge belongs to, but at the moment anyone can see the flag. 

Since it's a fair question though, I'll open a new issue as a question for that specific matter.

**Comment by iteles** [comment]
@asimonw The familiar user pattern here is that all users should be able to flag offensive comments to instil a sense of community moderation, which is why I believe when we tested this on our side we all assumed it was working as expected.

You raise an interesting point but I would suggest that we need a real use case to change this away from the 'default', raising a new issue on this specific question is a great idea ✨

**Comment by asimonw** [comment]
@iteles Agreed! For some reason, in internal conversations about this, we always referred to "the owner of the challenge," but you make a lot of sense :) #1033 created to follow up on this.

**Comment by asimonw** [comment]
@SimonLab One last issue I think. When as an admin I edit a flagged comment and click on "Save & Unflag" the UI doesn't get updated (the comment still looks flagged). After a page refresh the comment does appear unflagged, so it's only a UI issue.

If it doesn't require a lot of extra time, you can do the update with ajax. If it does, updating with a full refresh is fine for now, as it's an admin screen.
